test/toolkits/test_browser_toolkit.py

```diff
@@ -0,0 +1,198 @@
+# limitations under the License.
+# ========= Copyright 2023-2024 @ CAMEL-AI.org. All Rights Reserved. =========
+import io
+import os
+from typing import (
+    TYPE_CHECKING,
+)
+
+from PIL import Image
+
+if TYPE_CHECKING:
+    pass
+
+
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from camel.toolkits.browser_toolkit import (
+    AcceptedBrowserChannels,
+    BaseBrowser,
+    BrowserToolkit,
+)
+
+
+@pytest.fixture(scope="function")
+def base_browser_fixture():
+    with patch('playwright.sync_api.sync_playwright'):
+        browser = BaseBrowser(headless=True, cache_dir="test_cache")
+        yield browser
+        # Cleanup
+        if os.path.exists("test_cache"):
+            os.rmdir("test_cache")
+
+
+@pytest.fixture(scope="function")
+def browser_toolkit_fixture():
+    with patch('playwright.sync_api.sync_playwright'):
+        toolkit = BrowserToolkit(headless=True, cache_dir="test_cache")
+        yield toolkit
+        # Cleanup
+        if os.path.exists("test_cache"):
+            os.rmdir("test_cache")
+
+
+def test_base_browser_init(base_browser_fixture):
+    browser = base_browser_fixture
+    assert browser.headless is True
+    assert browser.cache_dir == "test_cache"
+    assert isinstance(browser.history, list)
+    assert isinstance(browser.page_history, list)
+
+
+def test_base_browser_initialization_order():
+    with patch('playwright.sync_api.sync_playwright'):
+        browser = BaseBrowser(headless=True)  # __init__ called
+        assert not hasattr(browser, 'browser')  # browser not launched yet
+        assert not hasattr(browser, 'page')  # page not created yet
+
+        browser.init()  # explicit init() call
+        assert hasattr(browser, 'browser')  # browser launched
+        assert hasattr(browser, 'page')  # page created
+
+
+def test_browser_chrome_channel_selection():
+    with patch('playwright.sync_api.sync_playwright'):
+        browser = BaseBrowser(
+            headless=True, channel=AcceptedBrowserChannels.CHROME
+        )
+        assert browser.channel == AcceptedBrowserChannels.CHROME.value
+
+
+def test_browser_msedge_channel_selection():
+    with patch('playwright.sync_api.sync_playwright'):
+        browser = BaseBrowser(
+            headless=True, channel=AcceptedBrowserChannels.MSEDGE
+        )
+        assert browser.channel == AcceptedBrowserChannels.MSEDGE.value
+
+
+def test_browser_chromium_channel_selection():
+    with patch('playwright.sync_api.sync_playwright'):
+        browser = BaseBrowser(
+            headless=True, channel=AcceptedBrowserChannels.CHROMIUM
+        )
+        assert browser.channel == AcceptedBrowserChannels.CHROMIUM.value
+
+
+def test_browser_visit_page(base_browser_fixture):
+    browser = base_browser_fixture
+    browser.init()
+
+    # Mock the page methods
+    browser.page.goto = MagicMock()
+    browser.page.wait_for_load_state = MagicMock()
+
+    test_url = "https://example.com"
```

Comment:
Define this outside the function so it can be reused by other functions
```python
TEST_URL = "..."
```

Reply:
Fixed.